File: 0-multer-single-file-upload/1-java/src/main/java/academy/starci/service/UploadService.java
package academy.starci.service;

import academy.starci.config.UploadProperties;
import academy.starci.dto.UploadedFileInfo;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

/**
 * Handles persisting the uploaded file to disk and building the response DTO.
 * Mirrors TS UploadService.describe(): filename = <epochMillis>-<originalName>.
 */
@Service
public class UploadService {

    private static final Logger logger = LoggerFactory.getLogger(UploadService.class);

    private final UploadProperties uploadProperties;

    public UploadService(UploadProperties uploadProperties) {
        this.uploadProperties = uploadProperties;
    }

    /**
     * Saves the multipart file to the configured destination directory and returns metadata.
     * Filename format: <System.currentTimeMillis()>-<originalFilename>
     *
     * @param file the validated multipart file
     * @return UploadedFileInfo matching the TS response shape
     * @throws IOException if writing to disk fails
     */
    public UploadedFileInfo save(MultipartFile file) throws IOException {
        // Ensure the upload directory exists — use absolute path so transferTo() works
        // regardless of the JVM working directory (Tomcat temp dir under Maven).
        Path destDir = Paths.get(uploadProperties.getDest()).toAbsolutePath();
        Files.createDirectories(destDir);

        // Build stored filename: <epochMillis>-<originalName>
        String originalName = file.getOriginalFilename() != null ? file.getOriginalFilename() : "upload";
        String storedFilename = System.currentTimeMillis() + "-" + originalName;
        Path storedPath = destDir.resolve(storedFilename);

        // Write bytes to disk.
        file.transferTo(storedPath.toFile());

        long sizeKb = Math.round(file.getSize() / 1024.0);
        logger.info("saved {} KB to {}", sizeKb, storedPath);

        // Normalise path separator to forward-slash for cross-platform consistency.
        String relPath = storedPath.toString().replace(File.separatorChar, '/');

        return new UploadedFileInfo(
                originalName,
                storedFilename,
                file.getSize(),
                file.getContentType(),
                relPath
        );
    }
}

File: 0-multer-single-file-upload/backend/0-typescript/src/modules/upload/upload.service.ts
import { Injectable, Logger } from "@nestjs/common"
import type { UploadedFileInfo } from "./upload.types"

/**
 * Service mô tả file đã upload — wrap metadata thành DTO phản hồi.
 * (EN: Service describes the uploaded file — wraps metadata into the response DTO.)
 */
@Injectable()
export class UploadService {
    private readonly logger = new Logger(UploadService.name)

    /**
     * Trả về metadata của file đã được Multer ghi xuống đĩa.
     * (EN: Returns metadata of the file already written to disk by Multer.)
     */
    public describe(file: Express.Multer.File): UploadedFileInfo {
        const sizeKb = Math.round(file.size / 1024)
        this.logger.log(`saved ${sizeKb} KB to ${file.path}`)
        return {
            originalName: file.originalname,
            filename: file.filename,
            size: file.size,
            mimetype: file.mimetype,
            path: file.path,
        }
    }
}
